### PR TITLE
Rebase #168

### PR DESCRIPTION
**Pull request by grahamas**
The main overwrite in the rebase was because of type promotion introduced in the master branch. For example, the function signature of multiplication is now `*(L::GhostDerivativeOperator{T1}, u::AbstractArray{T2}) where {T1,T2}` rather than `*(L::GhostDerivativeOperator{T,E,F}, u::AbstractArray{T,N}) where {T,E,F,N}`. Within each function, the initialization/allocation of the output is `x = similar(u, promote_type(T1,T2))`.  I didn't ensure that this pattern is consistent throughout.

The ensuing commits each fix some error/failure exposed by testing.

### CODE REVIEW

File: src/derivative_operators/bc_operators.jl [hunk 1]
```diff
@@ -1,47 +1,22 @@
...
+are not necessarily linear operators.  Instead, they are affine
+operators, with a constant term Q*x = Qa*x + Qb.
 """
 abstract type AffineBC{T} <: AtomicBC{T} end
 
 struct NeumannBC{N} end
 struct Neumann0BC{N} end
 struct DirichletBC{N} end
 struct Dirichlet0BC{N} end
-"""
```
Comment (xtalax) [inline src/derivative_operators/bc_operators.jl:15]: These docstrings shouldnt be being removed

File: test/multi_dim_bc_test.jl [hunk 2]
```diff
@@ -10,14 +10,14 @@ A = rand(n,m)
...
 #Create atomic BC
 q1 = RobinBC((1.0, 2.0, 3.0), (0.0, -1.0, 2.0), 0.1, 4.0)
-q2 = PeriodicBC(Float64)
+q2 = PeriodicBC{Float64}()
 
 BCx = vcat(fill(q1, div(m,2)), fill(q2, m-div(m,2)))  #The size of BCx has to be all size components *except* for x
 BCy = vcat(fill(q1, div(n,2)), fill(q2, n-div(n,2)))
 
 
-Qx = MultiDimBC{1}(BCx)
```
Comment (xtalax) [inline test/multi_dim_bc_test.jl]: This should be left as is to harmonise with the syntax for `CentredDifference`

File: test/multi_dim_bc_test.jl [hunk 3]
```diff
@@ -83,12 +88,15 @@ for N in 2:7
     sizes = rand(4:7, N)
     A = rand(sizes...)
 
-    Q1_N = Neumann0BC(Float64, Tuple(ones(N)), 3.0, size(A))
+    Q1_N = RobinBC(Tuple(rand(3)), Tuple(rand(3)), fill(0.1, N), 4.0, size(A))
```
Comment (grahamas) [inline test/multi_dim_bc_test.jl]: Is this syntax correct?  I don't think the size argument can go into the parameter

File: test/multi_dim_bc_test.jl (hunk 2 above)
Comment (grahamas) [reply test/multi_dim_bc_test.jl]: resolved by 61b0333, but see comment on line 91

File: src/derivative_operators/bc_operators.jl (hunk 1 above)
Comment (grahamas) [reply src/derivative_operators/bc_operators.jl:15]: Resolved by 5c14e31. However, I can't quite tell what the unicode should properly be. I copy-pasted it, but something may have been lost in translation.

File: test/multi_dim_bc_test.jl (hunk 3 above)
Comment (xtalax) [reply test/multi_dim_bc_test.jl]: this is the signature required for a multi dim bc, yeah

File: src/DiffEqOperators.jl [hunk 4]
```diff
@@ -22,25 +23,30 @@ include("utils.jl")
...
-include("derivative_operators/BC_operators.jl")
+include("derivative_operators/bc_operators.jl")
 include("derivative_operators/multi_dim_bc_operators.jl")
 
 ### Derivative Operators
 include("derivative_operators/fornberg.jl")
 include("derivative_operators/derivative_operator.jl")
 include("derivative_operators/abstract_operator_functions.jl")
 include("derivative_operators/convolutions.jl")
-include("derivative_operators/concretization.jl")
```
Comment (xtalax) [inline src/DiffEqOperators.jl]: Where did the concretizations go?

File: test/bc_coeff_compositions.jl [hunk 5]
```diff
@@ -216,25 +445,57 @@ end
...
+    As_l, As_b = sparse(A, length(u))
 
-    # Check ldiv!
-    u_temp = zeros(N)
-    ldiv!(u_temp, A, u)
-    @test u_temp ≈ ghost_u ≈ analytic_u
+    @test As_l ≈ analytic_AL #This is true
+    @test As_b ≈  analytic_Ab #This is true
+    @test Vector(u .- analytic_Ab) ≈ Vector(u .- As_b) #this is true
+    @test_broken analytic_AL\Vector(u .- analytic_Ab) ≈ As_l\Vector(u .- As_b) #Then why isn't this working?
```
Comment (xtalax) [inline test/bc_coeff_compositions.jl]: some help on this test would be fantastic - I can't understand why it isn't equivalent

File: src/DiffEqOperators.jl (hunk 4 above)
Comment (grahamas) [reply src/DiffEqOperators.jl]: The include moved down below (line 46). Not sure why.

File: test/bc_coeff_compositions.jl (hunk 5 above)
Comment (grahamas) [reply test/bc_coeff_compositions.jl]: uh oh, I screwed up this file.  I'll take a crack at reintegrating this week/weekend.

Comment (xtalax) [reply test/bc_coeff_compositions.jl]: I resolved some conflicts so pull before you continue

**Comment by grahamas** [comment]
I think I'm misunderstanding something about how this PR/rebase works -- at least some of the merge conflicts you resolved in the last commit were already handled in the rebase.  But now that I think about it, I'm not clear on how a PR/merge would deal with a rebase.  It makes sense that your attempted merge would still see those conflicts.  I think.

 See for example the changes in `src/DiffEqOperators.jl` in 3c11131. Those lines you introduced had been moved in the rebase to line up with where they are in the master branch.  Now there are duplicates.  It'll take me a bit more digging before I can figure out if similar things happened in the other files, since those are more involved changes (e.g. derivative_operators/concretization.jl, which has super-uninformative blobs). 

But before I dive in:  how is this rebase->PR supposed to work?

**Comment by xtalax** [comment]
I dont see any duplicates there - I am noticing that I resolved some of those docstring conflicts to the version with typos though. It may be that I have done something similar with other conflicts which is what is obfuscating how this works. I'll take a look at concretization.jl since i'm a bit more familiar with the super-uninformative blobs there lol.  

I have always done rebasing this way by pulling on to the active branch, resolving any conflicts to what should be based on the most modern version of the repo. Is there a better way?

**Comment by grahamas** [comment]
For duplicates, in `src/DiffEqOperators.jl` compare lines 9 and 13, as well as lines 46 and 53. 

[final edit: I think the below is actually unnecessary given my subsequent comment, so you can skip on down past this comment if you want.]

So here's what's happening, roughly:

1) I rebased, meaning I took the existing `SciML/DiffEqOperators` and then I applied all the commits from your branch on top of it.  Significantly, in some cases I overwrote your changes in favor of changes made on master since you branched off.
2) I PR'd from that rebased version onto your branch. My rebase and your branch now conflict because they have different commit histories.  I think this may have been a mistake? I think my rebase should probably only be merged directly onto `SciML/DiffEqOperators` since they share the same commit history trunk.
3) The final step is to PR onto `SciML/DiffEqOperators`. If properly rebased, this should amount to an automatic merge that purely applies new commits onto the tree.

So, now that I think about it, I think this PR is not the correct way forward.  For a proper rebase, the rebased branch should be directly merged onto the originating trunk (in this case, `SciML/DiffEqOperators`).  Otherwise we're just going to have to resolve the merge a third time when we eventually PR *this* branch onto `SciML/DiffEqOperators`.

Edit:  Actually, reviewing the commit log, I think this PR will work!  I *think* when this PR is accepted, the rebased commit history will be pulled through as well (at least, that's my interpretation of the commit history displayed at the top of this page; someone please confirm).  @xtalax I think you might want to undo your latest commit and go through the changes I made a bit more carefully -- I'm not sure the diffs on my changes adequately show why I made certain changes. In many cases, I overwrote your changes to be compatible with the current SciML master.  (though I *definitely* made mistakes, as we see in `test/bc_coeff_compositions.jl`)

@ChrisRackauckas do you have any experience with this kind of thing?  We should probably do it how you want, since you're the one who will have to approve the final PR

**Comment by grahamas** [comment]
I went through `concretization.jl` -- actually I think the only change made by the most recent commit was to delete a blank line at line 345.  Otherwise they're identical. 

https://github.com/xtalax/DiffEqOperators.jl/blame/c203185d5cde3a10f4ab4e44c6cfbf9e32b71c74/src/derivative_operators/concretization.jl

https://github.com/xtalax/DiffEqOperators.jl/blame/3c11131755ede5a1cdd702c35d88de37f471eb31/src/derivative_operators/concretization.jl

The first is without the most recent commit, the second is with the most recent commit.  Other than line 345, the only difference is the lack of blame information since, for whatever reason, the last commit's blobs got completely screwed up and acted like it overwrote the entire file.

So, in sum, *if possible* please remove the most recent commit, as it only changes white-space, adds duplicate lines, and eliminates line-by-line history in `concretization.jl`.  I think you can redo the merge, resolving all changes in favor of my branch (which might still nuke `concretization`'s blame history... sigh)

**Comment by grahamas** [comment]
An updated version was merged into DiffEqOperators

**Pull request closed without merging**